File: python/finished/macro.py
import traceback
import json
import re
import os


def isIPv6NetACL(cidr):
    if ":" in cidr:
        return True
    else:
        return False


def createEntry(isEgress, entry, resourceName, condition):
    splitset = entry.split(',')
    # Generate entry
    ruleNumber = splitset[0]
    protocol = splitset[1]
    action = splitset[2]
    cidr = splitset[3]
    if splitset[4] == "-1":
        fromPort = "-1"
        toPort = "-1"
    else:
        ports = splitset[4].split('-')
        fromPort = ports[0]
        toPort = ports[1] if len(ports) == 2 else ports[0]

    generatedEntry = {
        "Type": "AWS::EC2::NetworkAclEntry",
        "Properties": {
            "Egress": isEgress,
            "NetworkAclId": {
                "Ref": resourceName
            },
            "Protocol": protocol,
            "RuleAction": action,
            "RuleNumber": ruleNumber
        }
    }
    if condition:
        generatedEntry["Condition"] = condition

    if isIPv6NetACL(cidr):
        generatedEntry["Properties"]["Ipv6CidrBlock"] = cidr
    else:
        generatedEntry["Properties"]["CidrBlock"] = cidr
    if protocol != -1:
        generatedEntry["Properties"]["PortRange"] = {
            "From": fromPort,
            "To": toPort
        }

    entryType = 'Outbound' if isEgress else 'Inbound'

    # Generate name
    entryName = resourceName + entryType + ruleNumber
    return entryName, generatedEntry


def handler(event, context):
    macro_response = {
        "requestId": event["requestId"],
        "status": "success"
    }
    # Globals
    fragment = event['fragment']
    result = fragment

    for resource in list(fragment["Resources"].keys()):
        if "Condition" in fragment["Resources"][resource]:
            condition = fragment["Resources"][resource]["Condition"]
        else:
            condition = ''
        if fragment['Resources'][resource]['Type'] == 'AWS::EC2::NetworkAcl':
            if "Inbound" in fragment['Resources'][resource]['Properties']:
                for entry in fragment['Resources'][resource]['Properties']['Inbound']:
                    entryName, generatedEntry = createEntry(
                        False, entry, resource, condition)
                    result['Resources'][entryName] = generatedEntry
                result['Resources'][resource]['Properties'].pop('Inbound')
            if "Outbound" in fragment['Resources'][resource]['Properties']:
                for entry in fragment['Resources'][resource]['Properties']['Outbound']:
                    entryName, generatedEntry = createEntry(
                        True, entry, resource, condition)
                    result['Resources'][entryName] = generatedEntry
                result['Resources'][resource]['Properties'].pop('Outbound')
            if "Association" in fragment['Resources'][resource]['Properties']:
                for assoc in fragment['Resources'][resource]['Properties']['Association']:
                    generatedAssoc = {
                        "Type": "AWS::EC2::SubnetNetworkAclAssociation",
                        "Properties": {
                            "SubnetId": {"Ref": assoc},
                            "NetworkAclId": {"Ref": resource}
                        }
                    }
                    if condition:
                        generatedAssoc["Condition"] = condition
                    entryName = assoc + resource
                    result['Resources'][entryName] = generatedAssoc
                result['Resources'][resource]['Properties'].pop('Association')

    macro_response['fragment'] = result
    return macro_response

File: node/step3/test/macro.js
const macro = require('../macro');
var assert = require('assert');
var retError, result, fragment, requestId, generated_resource_name;

describe('handler', function () {
  context('Macro Returns Values', function () {
    before('Invoke function', function (done) {
      fragment = {
        "Resources": {
          "S3Bucket": {
              "Type": "AWS::S3::Bucket"
            }
          }
        };
      requestId = "testRequest";
      var event = {
        'region': 'ap-southeast-2',
        'requestId': requestId,
        'fragment': fragment
      };
      var context = {
        functionName: 'NaclExpander'
      };
      macro.handler(event, context, function (err, value) {
        retError = err ;
        result = value ;
        done();
      });
    });
    it('Check that success is returned', function () {
      assert.equal(result.status, "success");
    });
    it('Check that requestId is returned', function () {
      assert.equal(result.requestId, requestId);
    });
    it('Check input is passed through', function () {
      assert.equal(result.fragment, fragment);
    });
  });
  context('Nacl Association is created', function () {
    before('Invoke function', function (done) {
      generated_resource_name = 'SubnetANaclPublic';
      fragment = {"Resources": {"NaclPublic": { "Type": "AWS::EC2::NetworkAcl", "Properties": {"Association": ["SubnetA"]}}}};
      requestId = "testRequest";
      var event = {
        'region': 'ap-southeast-2',
        'requestId': requestId,
        'fragment': fragment
      };
      var context = {
        functionName: 'NaclExpander'
      };
      macro.handler(event, context, function (err, value) {
        retError = err ;
        result = value ;
        done();
      });
    });
    it('Check that success and requestId are returned', function () {
      assert.equal(result.status, "success");
    });
    it('Check that requestId is returned', function () {
      assert.equal(result.requestId, requestId);
    });
    it('Check all expected resources exist', function () {
      var expected_resources = ["NaclPublic", generated_resource_name]
      assert.deepEqual(Object.keys(result.fragment.Resources), expected_resources);
    });
    it('Check that correct type is set for association', function () {
      var expected_resource_type = 'AWS::EC2::SubnetNetworkAclAssociation';
      assert.equal(result.fragment.Resources[generated_resource_name].Type, expected_resource_type);
    });
    it('Check that correct properties are set for association', function () {
      var properties = result.fragment.Resources[generated_resource_name].Properties;
      assert.deepEqual(properties.NetworkAclId, {"Ref": "NaclPublic"});
      assert.deepEqual(properties.SubnetId, {"Ref": "SubnetA"});
    });
    it('Check that all Macro specific properties are removed', function () {
      assert.equal(result.fragment.Resources.NaclPublic.Properties.hasOwnProperty('Association'), false);
      assert.equal(result.fragment.Resources.NaclPublic.Properties.hasOwnProperty('Inbound'), false);
      assert.equal(result.fragment.Resources.NaclPublic.Properties.hasOwnProperty('Outbound'), false);
    });
  });
  context('Single Port Inbound', function () {
    before('Invoke function', function (done) {
      generated_resource_name = 'NaclPublicInbound100';
      fragment = {"Resources": { "NaclPublic": { "Type": "AWS::EC2::NetworkAcl", "Properties": { "Inbound": [ "100,6,allow,0.0.0.0/0,443"]}}}};
      requestId = "testRequest";
      var event = {
        'region': 'ap-southeast-2',
        'requestId': requestId,
        'fragment': fragment
      };
      var context = {
        functionName: 'NaclExpander'
      };
      macro.handler(event, context, function (err, value) {
        retError = err ;
        result = value ;
        done();
      });
    });
    it('Check that success and requestId are returned', function () {
      assert.equal(result.status, "success");
    });
    it('Check that requestId is returned', function () {
      assert.equal(result.requestId, requestId);
    });
    it('Check all expected resources exist', function () {
      var expected_resources = ["NaclPublic", generated_resource_name]
      assert.deepEqual(Object.keys(result.fragment.Resources), expected_resources);
    });
    it('Check that correct type is set for association', function () {
      var expected_resource_type = 'AWS::EC2::NetworkAclEntry';
      assert.equal(result.fragment.Resources[generated_resource_name].Type, expected_resource_type);
    });
    it('Check that correct properties are set for association', function () {
      var properties = result.fragment.Resources[generated_resource_name].Properties;
      assert.deepEqual(properties['CidrBlock'], '0.0.0.0/0');
      assert.deepEqual(properties['Protocol'], '6');
      assert.deepEqual(properties['Egress'], false);
      assert.deepEqual(properties['NetworkAclId'], { Ref: 'NaclPublic' });
      assert.deepEqual(properties['RuleAction'], 'allow');
      assert.deepEqual(properties['RuleNumber'], '100');
      assert.deepEqual(properties['PortRange']['From'], '443');
      assert.deepEqual(properties['PortRange']['To'], '443');
    });
    it('Check that all Macro specific properties are removed', function () {
      assert.equal(result.fragment.Resources.NaclPublic.Properties.hasOwnProperty('Association'), false);
      assert.equal(result.fragment.Resources.NaclPublic.Properties.hasOwnProperty('Inbound'), false);
      assert.equal(result.fragment.Resources.NaclPublic.Properties.hasOwnProperty('Outbound'), false);
    });
  });
  context('Single Port Outbound', function () {
    before('Invoke function', function (done) {
      generated_resource_name = 'NaclPublicOutbound100';
      fragment = {"Resources": { "NaclPublic": { "Type": "AWS::EC2::NetworkAcl", "Properties": { "Outbound": [ "100,6,allow,0.0.0.0/0,443"]}}}};
      requestId = "testRequest";
      var event = {
        'region': 'ap-southeast-2',
        'requestId': requestId,
        'fragment': fragment
      };
      var context = {
        functionName: 'NaclExpander'
      };
      macro.handler(event, context, function (err, value) {
        retError = err ;
        result = value ;
        done();
      });
    });
    it('Check that success and requestId are returned', function () {
      assert.equal(result.status, "success");
    });
    it('Check that requestId is returned', function () {
      assert.equal(result.requestId, requestId);
    });
    it('Check all expected resources exist', function () {
      var expected_resources = ["NaclPublic", generated_resource_name]
      assert.deepEqual(Object.keys(result.fragment.Resources), expected_resources);
    });
    it('Check that correct type is set for association', function () {
      var expected_resource_type = 'AWS::EC2::NetworkAclEntry';
      assert.equal(result.fragment.Resources[generated_resource_name].Type, expected_resource_type);
    });
    it('Check that correct properties are set for association', function () {
      var properties = result.fragment.Resources[generated_resource_name].Properties;
      assert.deepEqual(properties['CidrBlock'], '0.0.0.0/0');
      assert.deepEqual(properties['Protocol'], '6');
      assert.deepEqual(properties['Egress'], true);
      assert.deepEqual(properties['NetworkAclId'], { Ref: 'NaclPublic' });
      assert.deepEqual(properties['RuleAction'], 'allow');
      assert.deepEqual(properties['RuleNumber'], '100');
      assert.deepEqual(properties['PortRange']['From'], '443');
      assert.deepEqual(properties['PortRange']['To'], '443');
    });
    it('Check that all Macro specific properties are removed', function () {
      assert.equal(result.fragment.Resources.NaclPublic.Properties.hasOwnProperty('Association'), false);
      assert.equal(result.fragment.Resources.NaclPublic.Properties.hasOwnProperty('Inbound'), false);
      assert.equal(result.fragment.Resources.NaclPublic.Properties.hasOwnProperty('Outbound'), false);
    });
  });
});
